{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial for real-world data\n",
    "This tutorial provides step-by-step instructions for reproducing the real-world results presented in our paper. Before running the code, please ensure that you have downloaded the necessary data from our repository."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Domain clustering"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "import torch\n",
    "import sys\n",
    "import os\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "from SpaMV.spamv import SpaMV\n",
    "from SpaMV.utils import preprocess_dc, mclust\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "device = torch.device(\"cuda:1\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(device)\n",
    "dataname  ='6_Mouse_Embryo'\n",
    "data_rna = sc.read_h5ad('Data/' + dataname + '/adata_RNA.h5ad')\n",
    "data_peaks = sc.read_h5ad('Data/' + dataname + '/adata_peaks.h5ad')\n",
    "omics_names = ['Transcriptomics', 'Epigenomics']\n",
    "datasets = preprocess_dc([data_rna, data_peaks], omics_names)\n",
    "n_cluster = 14\n",
    "\n",
    "# wandb.init(project=dataname)\n",
    "# wandb.login()\n",
    "model = SpaMV(datasets, interpretable=False, omics_names=omics_names, device=device)\n",
    "model.train(dataname)\n",
    "# wandb.finish()\n",
    "\n",
    "datasets[0].obsm['SpaMV'] = model.get_embedding()\n",
    "mclust(datasets[0], n_clusters=n_cluster, key='SpaMV')\n",
    "fig, axes = plt.subplots(1, 3, figsize=(13.5, 4))  # 1 row, 2 columns\n",
    "\n",
    "sc.pp.neighbors(datasets[0], use_rep='SpaMV')\n",
    "sc.tl.umap(datasets[0])\n",
    "sc.pl.spatial(datasets[0], color='cluster', show=False, title='Annotation', ax=axes[0])\n",
    "sc.pl.umap(datasets[0], color='SpaMV', ax=axes[1], show=False, s=20, title='UMAP')\n",
    "sc.pl.spatial(datasets[0], color='SpaMV', show=False, title='SpaMV\\nARI: {:.3f}'.format(adjusted_rand_score(datasets[0].obs['cluster'], datasets[0].obs['SpaMV'])), ax=axes[2])\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interpretable dimension reduction"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "import torch\n",
    "import sys\n",
    "import os\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import math\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "from SpaMV.spamv import SpaMV\n",
    "from SpaMV.utils import plot_embedding_results, clr_normalize_each_cell, plot_topic_correlation_ratio_multimodal, plot_top_positive_correlations_boxplot\n",
    "import scanpy as sc\n",
    "import wandb\n",
    "import squidpy as sq\n",
    "\n",
    "def softmax(row):\n",
    "    shifted = row - np.max(row)  # Prevent overflow\n",
    "    exp_values = np.exp(shifted)\n",
    "    return exp_values / exp_values.sum()\n",
    "\n",
    "\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(device)\n",
    "for dataset in ['6_Mouse_Embryo']:\n",
    "    data_folder = 'Data/' + dataset\n",
    "    if dataset in ['4_Human_Lymph_Node', '9_Mouse_Thymus', '10_Mouse_Breast_Tumor']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_ADT.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Proteomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [10, 10]\n",
    "    elif dataset in ['6_Mouse_Embryo']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_ATAC.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Epigenomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [20, 5]\n",
    "    elif dataset in ['7_ME13_1']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_H3K27ac_ATAC.h5ad'), sc.read_h5ad(data_folder + '/adata_H3K27me3_ATAC.h5ad')]\n",
    "        omics_names = ['H3K27ac', 'H3K27me3']\n",
    "        zs_dim = 20\n",
    "        zp_dims = [5, 5]\n",
    "    elif dataset in ['11_ccRCC_Y7_T']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_MET.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Metabolomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [5, 5]\n",
    "    else:\n",
    "        raise Exception(f\"Unkonwn Dataset: {dataset}\")\n",
    "    obs_names = None\n",
    "    weights = []\n",
    "    betas = []\n",
    "    for i in range(len(data)):\n",
    "        data[i].var_names_make_unique()\n",
    "        data[i].X = data[i].X.astype(np.float32).toarray() if issparse(data[i].X) else data[i].X.astype(np.float32)\n",
    "        sc.pp.filter_cells(data[i], min_genes=math.ceil(data[i].n_vars / 100))\n",
    "        obs_names = data[i].obs_names if obs_names is None else obs_names.intersection(data[i].obs_names)\n",
    "        weights.append(10 if omics_names[i] == 'Proteomics' else 1)\n",
    "        betas.append(1)\n",
    "    data_ori = []\n",
    "    for i in range(len(data)):\n",
    "        data[i] = data[i][obs_names]\n",
    "        data_ori.append(data[i].copy())\n",
    "        sc.pp.filter_genes(data[i], min_cells=math.ceil(data[i].n_obs / 100))\n",
    "        if omics_names[i] == 'Transcriptomics':\n",
    "            data[i] = data[i][:, (data[i].X > 1).sum(0) > data[i].n_obs / 100]\n",
    "            sc.pp.highly_variable_genes(data[i], subset=False, n_top_genes=1000, flavor='seurat_v3')\n",
    "            sc.pp.normalize_total(data[i])\n",
    "            sc.pp.log1p(data[i])\n",
    "            data[i] = data[i][:, data[i].var_names[data[i].var.highly_variable]]\n",
    "        elif omics_names[i] == 'Proteomics':\n",
    "            data[i] = clr_normalize_each_cell(data[i])\n",
    "        elif omics_names[i] == 'Epigenomics':\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        elif omics_names[i] == 'Metabolomics':\n",
    "            sc.pp.normalize_total(data[i])\n",
    "            sc.pp.log1p(data[i])\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        else:\n",
    "            sc.pp.log1p(data[i])\n",
    "            sq.gr.spatial_neighbors(data[i])\n",
    "            sq.gr.spatial_autocorr(data[i], mode=\"moran\", genes=data[i].var_names, n_perms=100, n_jobs=1)\n",
    "            data[i] = data[i][:, data[i].uns['moranI'].index[data[i].uns['moranI']['pval_norm'] < .05]]\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        data_ori[i] = data_ori[i][:, data[i].var_names]\n",
    "    # wandb.init(project=dataset)\n",
    "    # wandb.login()\n",
    "    model = SpaMV(data, zs_dim=zs_dim, zp_dims=zp_dims, interpretable=True, weights=weights, betas=betas, omics_names=omics_names, device=device, random_seed=0)\n",
    "    model.train(dataset)\n",
    "    # wandb.finish()\n",
    "    z, w = model.get_embedding_and_feature_by_topic()\n",
    "    z = z.apply(softmax, axis=1)\n",
    "    plot_embedding_results(data, omics_names, z, w, show=True, save=False, size=100)\n",
    "\n",
    "    for i in range(len(data)):\n",
    "        plot_top_positive_correlations_boxplot(data[i], z, omics_names[i])\n",
    "\n",
    "    plot_topic_correlation_ratio_multimodal(data=data, omics_names=omics_names, z=z)\n",
    "    plt.show()\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Domain clustering for spatial H3K27ac-H3K27me3 mouse embryo"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import torch\n",
    "import sys\n",
    "import os\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "from SpaMV.spamv import SpaMV\n",
    "from SpaMV.utils import preprocess_dc, mclust\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "device = torch.device(\"cuda:1\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(device)\n",
    "dataname  ='7_ME13_1'\n",
    "data_ac = sc.read_h5ad('Data/' + dataname + '/adata_H3K27ac_ATAC.h5ad')\n",
    "data_me3 = sc.read_h5ad('Data/' + dataname + '/adata_H3K27me3_ATAC.h5ad')\n",
    "omics_names = ['H3K27ac', 'H3K27me3']\n",
    "datasets = preprocess_dc([data_ac, data_me3], omics_names)\n",
    "n_cluster = 21\n",
    "\n",
    "# wandb.init(project=dataname)\n",
    "# wandb.login()\n",
    "model = SpaMV(datasets, interpretable=False, omics_names=omics_names, device=device)\n",
    "model.train(dataname)\n",
    "# wandb.finish()\n",
    "\n",
    "datasets[0].obsm['SpaMV_old'] = model.get_embedding()\n",
    "mclust(datasets[0], n_clusters=n_cluster, key='SpaMV_old')\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 4))  # 1 row, 2 columns\n",
    "\n",
    "sc.pp.neighbors(datasets[0], use_rep='SpaMV_old')\n",
    "sc.tl.umap(datasets[0])\n",
    "sc.pl.umap(datasets[0], color='SpaMV_old', ax=axes[0], show=False, s=20, title='UMAP')\n",
    "sc.pl.spatial(datasets[0], color='SpaMV_old', show=False, ax=axes[1])\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "import torch\n",
    "import sys\n",
    "import os\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import math\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "from SpaMV_old.spamv import SpaMV\n",
    "from SpaMV_old.utils import plot_embedding_results, clr_normalize_each_cell, plot_topic_correlation_ratio_multimodal, plot_top_positive_correlations_boxplot\n",
    "import scanpy as sc\n",
    "import wandb\n",
    "import squidpy as sq\n",
    "\n",
    "def softmax(row):\n",
    "    shifted = row - np.max(row)  # Prevent overflow\n",
    "    exp_values = np.exp(shifted)\n",
    "    return exp_values / exp_values.sum()\n",
    "\n",
    "\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(device)\n",
    "for dataset in ['7_ME13_1']:\n",
    "    data_folder = 'Data/' + dataset\n",
    "    if dataset in ['4_Human_Lymph_Node', '9_Mouse_Thymus', '10_Mouse_Breast_Tumor']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_ADT.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Proteomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [10, 10]\n",
    "    elif dataset in ['6_Mouse_Embryo']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_ATAC.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Epigenomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [20, 5]\n",
    "    elif dataset in ['7_ME13_1']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_H3K27ac_ATAC.h5ad'), sc.read_h5ad(data_folder + '/adata_H3K27me3_ATAC.h5ad')]\n",
    "        omics_names = ['H3K27ac', 'H3K27me3']\n",
    "        zs_dim = 20\n",
    "        zp_dims = [5, 5]\n",
    "    elif dataset in ['11_ccRCC_Y7_T']:\n",
    "        data = [sc.read_h5ad(data_folder + '/adata_RNA.h5ad'), sc.read_h5ad(data_folder + '/adata_MET.h5ad')]\n",
    "        omics_names = ['Transcriptomics', 'Metabolomics']\n",
    "        zs_dim = 10\n",
    "        zp_dims = [5, 5]\n",
    "    else:\n",
    "        raise Exception(f\"Unkonwn Dataset: {dataset}\")\n",
    "    obs_names = None\n",
    "    weights = []\n",
    "    betas = []\n",
    "    for i in range(len(data)):\n",
    "        data[i].var_names_make_unique()\n",
    "        data[i].X = data[i].X.astype(np.float32).toarray() if issparse(data[i].X) else data[i].X.astype(np.float32)\n",
    "        sc.pp.filter_cells(data[i], min_genes=math.ceil(data[i].n_vars / 100))\n",
    "        obs_names = data[i].obs_names if obs_names is None else obs_names.intersection(data[i].obs_names)\n",
    "        weights.append(10 if omics_names[i] == 'Proteomics' else 1)\n",
    "        betas.append(1)\n",
    "    data_ori = []\n",
    "    for i in range(len(data)):\n",
    "        data[i] = data[i][obs_names]\n",
    "        data_ori.append(data[i].copy())\n",
    "        sc.pp.filter_genes(data[i], min_cells=math.ceil(data[i].n_obs / 100))\n",
    "        if omics_names[i] == 'Transcriptomics':\n",
    "            data[i] = data[i][:, (data[i].X > 1).sum(0) > data[i].n_obs / 100]\n",
    "            sc.pp.highly_variable_genes(data[i], subset=False, n_top_genes=1000, flavor='seurat_v3')\n",
    "            sc.pp.normalize_total(data[i])\n",
    "            sc.pp.log1p(data[i])\n",
    "            data[i] = data[i][:, data[i].var_names[data[i].var.highly_variable]]\n",
    "        elif omics_names[i] == 'Proteomics':\n",
    "            data[i] = clr_normalize_each_cell(data[i])\n",
    "        elif omics_names[i] == 'Epigenomics':\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        elif omics_names[i] == 'Metabolomics':\n",
    "            sc.pp.normalize_total(data[i])\n",
    "            sc.pp.log1p(data[i])\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        else:\n",
    "            sc.pp.log1p(data[i])\n",
    "            sq.gr.spatial_neighbors(data[i])\n",
    "            sq.gr.spatial_autocorr(data[i], mode=\"moran\", genes=data[i].var_names, n_perms=100, n_jobs=1)\n",
    "            data[i] = data[i][:, data[i].uns['moranI'].index[data[i].uns['moranI']['pval_norm'] < .05]]\n",
    "            sc.pp.highly_variable_genes(data[i], n_top_genes=1000, subset=True, flavor='seurat')\n",
    "        data_ori[i] = data_ori[i][:, data[i].var_names]\n",
    "    # wandb.init(project=dataset)\n",
    "    # wandb.login()\n",
    "    model = SpaMV(data, zs_dim=zs_dim, zp_dims=zp_dims, interpretable=True, weights=weights, betas=betas, omics_names=omics_names, device=device, random_seed=0)\n",
    "    model.train(dataset)\n",
    "    # wandb.finish()\n",
    "    z, w = model.get_embedding_and_feature_by_topic()\n",
    "    z = z.apply(softmax, axis=1)\n",
    "    plot_embedding_results(data, omics_names, z, w, show=True, save=False, size=100)\n",
    "\n",
    "    for i in range(len(data)):\n",
    "        plot_top_positive_correlations_boxplot(data[i], z, omics_names[i])\n",
    "\n",
    "    plot_topic_correlation_ratio_multimodal(data=data, omics_names=omics_names, z=z)\n",
    "    plt.show()\n"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SpaMV",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
